### ISSUE TITLE
Can't close websocket connection from client 

### ISSUE THREAD

**Issue by natasha4** [christophd/citrus#338]
Hi!
I am using Citrus 2.7.2 for testing web socket connections. In Citrus tests I create WebSocketClient to connect my Jetty server using this:
`CitrusEndpoints.websocket().client().requestUrl(String.format("%s%s", host, endpoint)).build()`

I would like to create web socket connection before each scenario and close it after. But I didn't find close method. Is it possible to close socket from client?

Thank you!

**Comment by FMAOuroboros** [comment]
Hi!

So basically, Citrus does not directly manage the websocket session life cycle. Citrus implements the required interfaces to leverage the Spring implementation.

The Builder, generating the websocket client, does not open the session directly.
```java
CitrusEndpoints
    .websocket()
    .client()
    .requestUrl(String.format("%s%s", host, endpoint))
    .build()
```
The session is created, when you send your first message in your test.

What you could try to enforce new sessions, is to build a new websocket client before you run your test cases.
Something like:
```java
private WebSocketClient webSocketClient;

@BeforeMethod
public void newClient(){
    webSocketClient = CitrusEndpoints
            .websocket()
            .client()
            .requestUrl(String.format("%s%s", host, endpoint))
            .build();
}
```
I haven't tested it locally but this should work. I hope that helps you out.

It would be brilliant if you could give me just a few more details on your test scenario and your use case for opening and closing sessions so that I can help you out with more detailed information. Especially the use case for manually closing the sessions would be interesting.

**Comment by natasha4** [comment]
Hi @FMAOuroboros !

Thanks a lot for your detailed answer!

I do exactly what you suggested: building a new websocket client in @Before method. So I have a new session for every test case.

The reason why I want to open and close sessions:
In some test cases I authorise user, in others I don't need any user to be authorised. So to make all test cases independent I need  a new web socket session for every test case. So I open it before each test case. 
The next thing: I want to close them because I don't want old sessions hang. Also after some time they start to close by Idle Timeout and I see error stacktraces in logs. (Actually it will happen only if my tests will be executed longer than 5 mins. Right now I don't have a lot of tests and the suit finishes before I see these stacktraces)
I think that conceptually open session in @Before method and close it in @After method is right.

I thought may be Citrus could expose somehow session for using this in a client code. Something like this:

```java
WebSocketClient client = CitrusEndpoints.websocket()
                .client()
                .requestUrl(requestUrl)
                .build();

 CitrusWebSocketHandler handler = new CitrusWebSocketHandler() {
            @Override
            public void afterConnectionEstablished(WebSocketSession session) throws Exception {
                     //here it will be possible to save session to some variable in my code    
                     super.afterConnectionEstablished(session);
            }
};

 client.getEndpointConfiguration().setHandler(handler);
```
Currently `com.consol.citrus.websocket.client.WebSocketClientEndpointConfiguration.setHandler()` throws `UnsupportedOperationException`.

**Comment by FMAOuroboros** [comment]
Hi!

Glad that the workaround helps you out!

Thanks a lot for the details! Makes sense. I don't know your application and how it works but what comes to my mind is, couldn't you setup the session as you need it per test?
```java
WebSocketClient client = CitrusEndpoints.websocket()
                .client()
                .requestUrl(requestUrl)
                .build();
@Test
@CitrusTest
public void testSomethingWithAuthorizedUser() {
    loginUser(client)

    performTestWithAuthorizedUser()
    
    logoutUser(client)
}

@Test
@CitrusTest
public void testSomethingWithUnauthorizedUser() {
    performTestWithUnauthorizedUser()
}
```

I mean basically a new session doesn't seem to be required, but a session is a certain state, so that the test works as expected. You just have to follow the convention that there is no authorized user before the test and no authorized user after.
If this assumption is incorrect for your use case, please let me know!

Nevertheless, leaving sessions open is per se nothing you should do in productive systems but because citrus just runs in a test environment that is shutdown after the tests have been finished, there are not really dangling sessions. The stack traces caused by the timeouts on the other hand could cause troubles, because they look like an error but are technically not relevant to the test. Therefore I opened #346 to keep track of that.

Concerning your idea of implementing a `CitrusWebSocketHandler`:
Basically a great idea! I opened #347 to open citrus for this use case.

BR,
Sven

**Comment by natasha4** [comment]
Hi @FMAOuroboros !

Thank you for your advice about user logout! We had such thought. But we decided that having clean session is more easier for us than thinking about returning session to the initial state after test execution.

So now I use Java Reflection to get session object and close it. It is not good approach at all, but that is fine for us now. 

Thank you,
Natalia

**Comment by FMAOuroboros** [comment]
Hi @natasha4!

Good to hear that you solved your problem. I hope, we get #347 solved soon. Than you'll be able to remove the reflections and work with the sessions directly.

BR,
Sven
